lib/Sema/CSGen.cpp

```diff
@@ -356,7 +389,183 @@ namespace {
+  }
+
+  bool expandCollectedTypesFromOperatorOverloads(ConstraintSystem &CS,
+                                                 LinkedTypeInfo &lti) {
+    while (1) {
+      llvm::SmallSet<TypeBase *, 4> additionalTypes;
+
+      for (auto opExpr : lti.operatorExprs) {
+        auto literalArgs = literalArgsForExpr(opExpr);
+        auto fn = opExpr->getFn();
+        ArrayRef<ValueDecl *> decls;
+
+        if (auto odre = dyn_cast<OverloadedDeclRefExpr>(fn))
+          decls = odre->getDecls();
+        else if (auto le = dyn_cast<LookupExpr>(fn))
+          decls = ArrayRef<ValueDecl *>(le->getMember().getDecl());
+        else
+          continue;
+
+        for (auto decl : decls) {
+          auto metaFnType = decl->getInterfaceType()->getAs<AnyFunctionType>();
+          auto fnType = metaFnType->getResult()->getAs<AnyFunctionType>();
+          if (!fnType)
+            fnType = metaFnType;
+          auto resultTy =
+              fnType->getResult()->getRValueType()->getDesugaredType();
+
+          if (lti.collectedTypes.count(resultTy) > 0)
+            continue;
+
+          bool possibleDecl = true;
+
+          if (auto generic = resultTy->getAs<GenericTypeParamType>()) {
+            bool hasEquivalentParam = false;
+            for (auto param : fnType->getParams())
+              hasEquivalentParam |= resultTy->isEqual(param.getPlainType());
+            if (hasEquivalentParam)
+              continue;
+          }
+
+          auto literalArgIter = literalArgs.begin();
+          for (auto param : fnType->getParams()) {
+            auto paramTy =
+                param.getPlainType()->getRValueType()->getDesugaredType();
+            if (paramTy->is<DependentMemberType>() ||
+                paramTy->is<BoundGenericEnumType>() || param.isAutoClosure())
+              return false;
+
+            if (!isTypeInCollectedTypes(CS, lti, paramTy, *literalArgIter)) {
+              possibleDecl = false;
+              break;
+            }
+
+            // If isTypeInCollectedTypes but not literally in the set, it's
+            // convertibleTo, add it.
+            if (!lti.collectedTypes.count(paramTy) && !*literalArgIter &&
+                !paramTy->is<GenericTypeParamType>())
+              additionalTypes.insert(paramTy);
+
+            literalArgIter++;
+          }
+          if (possibleDecl)
+            additionalTypes.insert(resultTy);
+        }
+      }
+
+      if (additionalTypes.size() == 0)
+        return true;
+
+      for (auto type : additionalTypes)
+        lti.collectedTypes.insert(type);
+    }
+  }
+
+  void disableUnreachableOperatorOverloads(ConstraintSystem &CS,
+                                           LinkedTypeInfo &lti) {
+    for (auto opExpr : lti.operatorExprs) {
+      auto tyvar = CS.getType(opExpr->getFn())->getAs<TypeVariableType>();
+      if (!tyvar)
+        continue;
+
+      llvm::TinyPtrVector<Constraint *> disjunctions =
+          CS.getConstraintGraph().gatherConstraints(
```

Comment:
Performance wise - it would be better to invert that check and have outer loop be for every disjunction constraint, at least it's going to be linear over a number of constraints in the system.